### ISSUE TITLE
What is the expected output for test_client / test_server?

### ISSUE THREAD

**Issue by op8867555** [fcitx/xcb-imdkit#2]
Hi there, I'm new to XCB/XIM and I'm trying to create a XCB window application with XIM supporting.
I have tried running `test_client` and `test_server` (starting from running server, with `XIM_DEBUG=1`) at the same time, which gave me:

```
> ./test/test_client
-- XIM_REGISTER_TRIGGERKEYS
-- XIM_OPEN_REPLY
-- XIM_QUERY_EXTENSION_REPLY
-- XIM_ENCODING_NEGOTIATION_REPLY
-- XIM_GET_IM_VALUES_REPLY
-- XIM_CREATE_IC_REPLY
icid:1
-- XIM_SET_IC_VALUES_REPLY
set ic 1 done
-- XIM_GET_IC_VALUES_REPLY
get ic 1 done
-- XIM_FORWARD_EVENT
Key press Keycode 65, State 4
```

```
> ./test/test_server
winid:31457280
requestor:52428800
buf:@transport=X/
-- XIM_CONNECT
-- XIM_OPEN
-- XIM_QUERY_EXTENSION
-- XIM_ENCODING_NEGOTIATION
-- XIM_GET_IM_VALUES
-- XIM_CREATE_IC
-- XIM_SET_IC_VALUES
-- XIM_GET_IC_VALUES
-- XIM_FORWARD_EVENT
```

I'm assuming `test_client` should show me more events, but it didn't look like that.

**Comment by wengxt** [comment]
Well, it's not really a automatic test, it is mostly for me to verify the things manually. (Because you'll need an X Server to run the test, I didn't yet spend time to write a fully automated test).

The test_server is a simple xim server, that will by-pass every key event you send to it, except when you press 't', it will commit a hard-coded string to the client.

The test_client simulates the client and manually forward some key events.

If you're implement a XCB client and want to add xim support, you can take a look at main() in test_client and use test_server to test against it.

Basically the idea is to first setup the xim connection as in main() and find the location where you handle the XCB event and call xcb_xim_filter_event before you handle the event. The return value of xcb_xim_filter_event will tell you whether this event is handled by XIM protocol or not.

One common mistake is that only pass xkb_key_press and xkb_key_release to xim, which is wrong because there are a lot of client message used in xim. You should pass every event you received to xcb_xim_filter_event.

Hope this helps.

**Comment by op8867555** [comment]
Thanks for the reply! I still have some questions below.

> call xcb_xim_filter_event before you handle the event

Do I have to handle trigger_keys (to send a trigger notify) or forward key press events by myself?

> xcb_xim_filter_event will tell you whether this event is handled by XIM protocol or not.

Does this mean I can just skip events which handled by XIM?

In the case it is not clear enough, here is a minimal example (not working) I'm working on:
https://gist.github.com/op8867555/85e8adac89142f2cbad02c2ca3c0033a

**Comment by johnbeard** [comment]
@wengxt I also have the same question:
> Do I have to handle trigger_keys (to send a trigger notify) or forward key press events by myself?

Basically, where should you be calling `xcb_xim_forward_event` from in a real program?

And do you expect to see `XIM_FORWARD_EVENT` more than once from the test_client? `xcb_xim_forward_event` is called many times in `get_ic_values_callback`but only one `XIM_FORWARD_EVENT` is reported.

**Comment by wengxt** [comment]
@johnbeard, the main loop in my mind should looks like this:

Despite of the setup part:
```
    xcb_generic_event_t *event;
    while ((event = xcb_wait_for_event(connection))) {
        if (xcb_xim_filter_event(im, event)) {
              free(event);
              continue;
        }
        if ((event->response_type & ~0x80) == XCB_KEY_PRESS || (event->response_type & ~0x80) == XCB_KEY_RELEASE) && we_can_do_text_input) {
             xcb_xim_forward_event(im, ic, event);
        }
        free(event);
    }

```

we_can_do_text_input ( means something about your application logic, e.g. you know whether your application has focus in a text field or not)

Hope this helps. The test_client is just simulating keyboard event and the "callback" are just used to chain the test together, it doesn't reflect anything in the real application.

**Comment by tlefebvr** [comment]
Hi @wengxt,

Sorry to revive this issue, but the loop you wrote doesn't change anything for me : only one `XIM_FORWARD_EVENT` reply and no other key events are sent to the server, whether it be fcitx or your test server.

I have tried to debug the test and observed that messages created by `xcb_xim_forward_event` are queuing after the first one instead of being sent to the IM server, since the loop in `_xcb_xim_process_queue` is never entered after this one.
I tried a dirty hack to solve that, clearing the current waiting message and processing the queue at the end of `_xcb_xim_handle_forward_event`, but the library never executes that code since it doesn't receive forward events from the server anymore, so it's also possible that I have misunderstood IM servers' behaviour.

Does your test really give you the results you expect from your side ? Is this definitely a bug ? Or does the main loop not handle every replies correctly ?
